service/controller/resource/workermigration/create.go

```diff
@@ -0,0 +1,496 @@
+		return microerror.Mask(err)
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", "ensured that built-in workers have drainerconfig cr")
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "ensuring that built-in workers are drained")
+	allNodesDrained, err := r.allDrainerConfigsWithDrainedState(ctx, cr)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if !allNodesDrained {
+		r.logger.LogCtx(ctx, "level", "debug", "message", "some old worker nodes are still draining")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "deleting timed out drainerconfigs")
+
+		// In case some draining operations timed out, delete CRs so that they
+		// are recreated on next reconciliation.
+		err = r.deleteTimedOutDrainerConfigs(ctx, cr)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		r.logger.LogCtx(ctx, "level", "debug", "message", "deleted timed out drainerconfigs")
+		r.logger.LogCtx(ctx, "level", "debug", "message", "canceling resource")
+
+		return nil
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", "ensured that built-in workers are drained")
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "deleting drainerconfigs")
+	err = r.deleteDrainerConfigs(ctx, cr)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", "deleted drainerconfigs")
+
+	r.logger.LogCtx(ctx, "level", "debug", "message", "deleting built-in workers' VMSS")
+	err = azureAPI.DeleteVMSS(ctx, key.ResourceGroupName(cr), *builtinVMSS.Name)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	r.logger.LogCtx(ctx, "level", "debug", "message", "deleted built-in workers' VMSS")
+
+	return nil
+}
+
+func (r *Resource) allDrainerConfigsWithDrainedState(ctx context.Context, cr providerv1alpha1.AzureConfig) (bool, error) {
+	o := client.MatchingLabels{
+		capiv1alpha3.ClusterLabelName: key.ClusterID(&cr),
+	}
+
+	var dcList corev1alpha1.DrainerConfigList
+	err := r.ctrlClient.List(ctx, &dcList, o)
+	if err != nil {
+		return false, microerror.Mask(err)
+	}
+
+	for _, dc := range dcList.Items {
+		if !dc.Status.HasDrainedCondition() {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func (r *Resource) deleteTimedOutDrainerConfigs(ctx context.Context, cr providerv1alpha1.AzureConfig) error {
+	o := client.MatchingLabels{
+		capiv1alpha3.ClusterLabelName: key.ClusterID(&cr),
+	}
+
+	var dcList corev1alpha1.DrainerConfigList
+	err := r.ctrlClient.List(ctx, &dcList, o)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, dc := range dcList.Items {
+		if dc.Status.HasTimeoutCondition() {
+			err = r.ctrlClient.Delete(ctx, &dc)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *Resource) deleteDrainerConfigs(ctx context.Context, cr providerv1alpha1.AzureConfig) error {
+	o := client.MatchingLabels{
+		capiv1alpha3.ClusterLabelName: key.ClusterID(&cr),
+	}
+
+	var dcList corev1alpha1.DrainerConfigList
+	err := r.ctrlClient.List(ctx, &dcList, o)
```

Comment:
could we just delete all objects that match? Like here https://github.com/kubernetes-sigs/controller-runtime/blob/5757a389803ec368126bb1ff046ae3524dacbfcf/pkg/client/example_test.go#L238